ENH: Use AVX for float32 implementation of np.exp & np.log

r-devulap opened the pull request:
This commit implements vectorized single precision exponential and
natural log using AVX2 and AVX512.

Accuracy:

| Function | Max ULP Error | Max Relative Error |
|----------|---------------|--------------------|
| np.exp   | 2.52          | 2.1E-07            |
| np.log   | 3.83          | 2.4E-07            |

Performance:

(1) Micro-benchmarks: measured execution time of np.exp and np.log using
timeit package in python. Each function is executed 1000 times and this
is repeated 100 times. The standard deviation for all the runs was less
than 2% of their mean value and hence not included in the data. The
vectorized implementation was upto 7.6x faster than the scalar version.

| Function | NumPy1.16 | AVX2   | AVX512 | AVX2 speedup | AVX512 speedup |
| -------- | --------- | ------ | ------ | ------------ | -------------- |
| np.exp   | 0.395s    | 0.112s | 0.055s | 3.56x        | 7.25x          |
| np.log   | 0.456s    | 0.147s | 0.059s | 3.10x        | 7.64x          |

(2) Logistic regression: exp and log are heavily used in training neural
networks (as part of sigmoid activation function and loss function
respectively). This patch significantly speeds up training a logistic
regression model. As an example, we measured how much time it takes to
train a model with 15 features using 1000 training data points. We
observed a 2x speed up to train the model to achieve a loss function
error < 10E-04.

| Function       | NumPy1.16  | AVX2   | AVX512 | AVX2 speedup | AVX512 speedup |
| -------------- | ---------- | ------ | ------ | ------------ | -------------- |
| logistic.train | 121.0s     | 75.02s | 60.60s | 1.61x        | 2.02x          |

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html#writing-the-commit-message
-->

Commit:
ENH: vectorizing float32 implementation of np.exp & np.log

This commit implements vectorized single precision exponential and
natural log using AVX2 and AVX512.

Accuracy:

| Function | Max ULP Error | Max Relative Error |
|----------|---------------|--------------------|
| np.exp   | 2.52          | 2.1E-07            |
| np.log   | 3.83          | 2.4E-07            |

Performance:

(1) Micro-benchmarks: measured execution time of np.exp and np.log using
timeit package in python. Each function is executed 1000 times and this
is repeated 100 times. The standard deviation for all the runs was less
than 2% of their mean value and hence not included in the data. The
vectorized implementation was upto 7.6x faster than the scalar version.

| Function | NumPy1.16 | AVX2   | AVX512 | AVX2 speedup | AVX512 speedup |
| -------- | --------- | ------ | ------ | ------------ | -------------- |
| np.exp   | 0.395s    | 0.112s | 0.055s | 3.56x        | 7.25x          |
| np.log   | 0.456s    | 0.147s | 0.059s | 3.10x        | 7.64x          |

(2) Logistic regression: exp and log are heavily used in training neural
networks (as part of sigmoid activation function and loss function
respectively). This patch significantly speeds up training a logistic
regression model. As an example, we measured how much time it takes to
train a model with 15 features using 1000 training data points. We
observed a 2x speed up to train the model to achieve a loss function
error < 10E-04.

| Function       | NumPy1.16  | AVX2   | AVX512 | AVX2 speedup | AVX512 speedup |
| -------------- | ---------- | ------ | ------ | ------------ | -------------- |
| logistic.train | 121.0s     | 75.02s | 60.60s | 1.61x        | 2.02x          |

juliantaylor commented on numpy/core/src/umath/simd.inc.src:
```diff
@@ -1075,6 +1096,386 @@ sse2_@kind@_@TYPE@(@type@ * ip, @type@ * op, const npy_intp n)
+NPY_INLINE NPY_GCC_OPT_3 NPY_GCC_TARGET_AVX2 __m256
+avx2_fmadd(__m256 a, __m256 b, __m256 c)
+{
+    return _mm256_add_ps(_mm256_mul_ps(a, b), c);
```
avx2 has fmadd natively, does it not have the required semantics for this and avx512 fmadd does?

r-devulap replied:
It does, but it additionally requires "fma" attribute. FMA was introduced post AVX2, so it wouldn't work with the older CPU's which have AVX2 but no FMA.

juliantaylor replied:
hm right
Though I wonder if this has an effect on the accuracy bounds you quoted for avx2

r-devulap replied:
The max ULP error and relative error reported above actually holds true for both AVX2 and AVX512 implementations. They were computed by comparing the output of these functions to glibc's scalar implementation of exp and log in double precision (higher precision) across all 32-bit floats (2^32 values). I did that for both AVX2 and AVX512 versions and turns out the max error is the same for both of them.

juliantaylor replied:
But the two variants do probably not return the exact same results?
For reproducibility it would be preferable to minimize the amount of different results on different hardware generations.
Only supporting this for AVX2+fma hardware is probably not unreasonable.

Though do you know how common avx2 hardware without fma is?

r-devulap replied:
So, I did run an exhaustive check for all float32 and I verified that avx2/avx512 version of log and exp do produce the exact same result (0 ULP error, 0.0 absolute error). 

Intel introduced FMA in Haswell. So, Ivy Bridge and Sandy Bridge (released in 2011-2012) don't have FMA's. Since they seem to be producing the same result, I would like to keep it as is if that is okay with you. 

> very nice, can you provide the code you used to compare the implementations?
> 
> please do not add sin/cos to this PR, lets do that one at the time.

juliantaylor commented:
This looks very nice and uses our runtime detection so automatic performance improvement for users with the right hardware using gcc.

It is always a bit dodgy to implement the basic functions ourselves and not using libm for it as we lose out on potential improvements there (e.g. glibc has avx512 code for the these functions and more).
Then again we gain more control over when they change and are more reliable over different platforms.

Would be interesting to test this against a few other projects like scipy and see if tests break.

r-devulap commented:
> This looks very nice and uses our runtime detection so automatic performance improvement for users with the right hardware using gcc.
> 
> It is always a bit dodgy to implement the basic functions ourselves and not using libm for it as we lose out on potential improvements there (e.g. glibc has avx512 code for the these functions and more).
> Then again we gain more control over when they change and are more reliable over different platforms.
> 
> Would be interesting to test this against a few other projects like scipy and see if tests break.

Thank you for reviewing this patch! I did run the test suite in SciPy with this version of NumPy on my SkylakeX machine and didn't see any tests fail.

juliantaylor commented:
Do you have a reference for the algorithm used here?

juliantaylor commented:
I would like to compare this glibc's exp implementation (there is an old branch of mine which allows using it from numpy https://github.com/numpy/numpy/pull/7865) but mostly just out of curiosity.

A concern is that we do not have good coverage if these basic math functions actually compute correct results over the full range, while I trust it is correct currently to keep it correct coverage would be good.
Then again the same thing applies to our fallback implementations in case the system libm does not provide decent functions... so it is not any worse than the status quo if we merge as is.

juliantaylor commented on numpy/core/src/umath/simd.inc.src:
```diff
@@ -1075,6 +1096,386 @@ sse2_@kind@_@TYPE@(@type@ * ip, @type@ * op, const npy_intp n)
+}
+
+NPY_INLINE NPY_GCC_OPT_3 NPY_GCC_TARGET_AVX2 __m256
+avx2_get_partial_load_mask(const npy_int num_elem, const npy_int total_elem)
```
the functions all need to be static as we do not need external linkage for them

r-devulap commented:
> Do you have a reference for the algorithm used here?

The algorithm is based of the book "Elementary Functions: Algorithms and Implementation" by Jean-Michel Muller. Basically algorithm for transcendental functions is broken into 3 steps: 

1. Range reduction (chapter 9 talks about this for sin, cos, exp and log). I ended up using the Cody-Waite range reduction method which is very SIMD friendly. 
2. Polynomial approximation in the reduced range (using a mini-max polynomial approximation), the coefficients of which can be computed by solving Remez's algorithm. [Boost library](https://www.boost.org/doc/libs/1_67_0/libs/math/doc/html/math_toolkit/remez.html) provides a nice solver for this and I used it to experiment with different polynomials and used one that provided reasonable accuracy.  
3. Reconstruction of the original result, which is a simple multiplication or addition in this case.

r-devulap commented:
> I would like to compare this glibc's exp implementation (there is an old branch of mine which allows using it from numpy #7865) but mostly just out of curiosity.
> 
> A concern is that we do not have good coverage if these basic math functions actually compute correct results over the full range, while I trust it is correct currently to keep it correct coverage would be good.
> Then again the same thing applies to our fallback implementations in case the system libm does not provide decent functions... so it is not any worse than the status quo if we merge as is.

I did compare the accuracy of this implementation with glibc's libmvec library (32-bit version) and turns out libmvec has a slightly worse accuracy that this implementation. 

| Function | libmvec max ULP | this patch max ULP | 
| -- | -- | -- |
| sine | 2.47 | 1.49 |
| cosine | 2.39 | 1.49 |
| exp | 2.93 | 2.52 |
| log | 3.96 | 3.83 |

I have another patch ready which implements float32 implementation of sine and cosine using AVX2 and AXV512 with a max ULP of 1.49. If you like, I can add that to this pull request as well.

juliantaylor commented:
very nice, can you provide the code you used to compare the implementations?

please do not add sin/cos to this PR, lets do that one at the time.

r-devulap commented:
> very nice, can you provide the code you used to compare the implementations?
> 
> please do not add sin/cos to this PR, lets do that one at the time.

The code is fairly simple actually: 

Have a for loop iterating over all 32 bit floats and for each float: 

1. `double my_value = (double) AVX2_exp_FLOAT(); // upscale the float to a double`
2. `double reference_value = exp(); // glibc's scalar function `
3. `double ULP_error  = compute_ULP(my_value, reference_value); // a small helper function which computes the error is terms of ULP`  

you can then compute the max ULP error and do the same to get max relative error.

Commit:
BUG: Fixing compile time error for clang

1) Got rid of @isa@_cmp_mask helper function, since clang expects the
compare operator in _mm@vsize@_cmp_ps to be a compile time constant

2) Converted all helper functions to static functions

mattip commented:
There are warnings when compiling with various options, are some of the #ifdef clauses not quite correct?

r-devulap commented:
> There are warnings when compiling with various options, are some of the #ifdef clauses not quite correct?

Could you please tell me which compiler and flags you are using? With gcc 7.3.0, it seems to build fine. There are some warnings, but nothing seems to be from this patch. 

If you are using clang6.0.0, then this patch fails to build because of this error (gcc or even clang7.0 builds fine): 

_configtest.c:7:20: error: instruction requires: AVX-512 ISA                                                                                                                                  
`  __asm__ volatile("vpaddd %zmm1, %zmm2, %zmm3");`

mattip commented:
The CI is failing. You should see a red X next to the changeset label above:
```
BUG: Fixing compile time error for clang                               X f802e92 
```
Clicking on that X should open a set of links to failing CI runs: [travis ci](https://travis-ci.org/numpy/numpy/builds/510032107?utm_source=github_status&utm_medium=notification),  and [azure macos and 32 bit linux](https://dev.azure.com/numpy/numpy/_build/results?buildId=2556&view=logs). The CI fails on compiler warnings.

Commit:
BUG: Fixing AVX512F build issues on clang6.0

clang6.0 fails to compile this code:

__asm__ __volatile__ (
    "vpaddd %zmm1, %zmm2, %zmm3\n\t"
);

Note that this is a known issue in clang6.0. clang7.0 and gcc does not
have this problem. This fails to set the flag HAVE_LINK_AVX512F. Hence,
the AVX512F version of exp and log doesn't get built. If AVX512F is
detected during runtime, instead of choosing to run the AVX2 version, it
will end up running scalar version.

mattip commented:
Linux 32 bit is [failing](https://dev.azure.com/numpy/numpy/_build/results?buildId=2607&view=logs&jobId=1cdee7e3-4100-5086-4c59-ed0352998504) - `ImportError: /numpy/build/testenv/lib/python3.6/site-packages/numpy/core/_multiarray_umath.cpython-36m-i386-linux-gnu.so: undefined symbol: AVX512F_exp_FLOAT`

r-devulap commented:
> Linux 32 bit is [failing](https://dev.azure.com/numpy/numpy/_build/results?buildId=2607&view=logs&jobId=1cdee7e3-4100-5086-4c59-ed0352998504) - `ImportError: /numpy/build/testenv/lib/python3.6/site-packages/numpy/core/_multiarray_umath.cpython-36m-i386-linux-gnu.so: undefined symbol: AVX512F_exp_FLOAT`

yeah, I still don't fully understand why that is failing. Do you have any suggestions?

tylerjereddy commented:
Is it worth trying `-march=skylake-avx512` or similar compile flag for the 32-bit Ubuntu docker container? Maybe 32-bit linux requires a little more convincing to emit the appropriate code / use appropriate libs on that platform?

tylerjereddy commented:
Even if that did work, I guess you'd still need some kind of guard so that you can still import NumPy on a "regular" compile without the -m flag on 32-bit.

Other random thought--that Docker container will likely have pretty minimal libs installed so if you're expecting something specific it could be missing.

r-devulap commented:
hmm, I don't have any dependencies on other libs. If GCC supports avx512f instrinsics, then it compiles it. At the time of loading .so file (when importing numpy), if cpuid/xgetbv returns a successful check for avx512f, then it points to the avx512f version of the function FLOAT_exp_avx512f, which is where the failure seems to occur.

mattip commented:
fired up my bionic 32 bit chroot. Forcing the archtecture with ``CFLAGS='-march=skylake-avx512'` indeed allows the module to load, but then it crashes with `Illegal instruction` when running tests. Apparently the check is not quite good enough for a chroot.

mattip commented:
We have had reports of ["illegal instruction"](https://github.com/numpy/numpy/issues?utf8=%E2%9C%93&q=is%3Aissue+%22illegal+instruction%22+) in the past. Probably the best course is check for one of the [`__SSE__` defines](https://stackoverflow.com/questions/28939652/how-to-detect-sse-sse2-avx-avx2-avx-512-avx-128-fma-kcvi-availability-at-compile) on the build platform, and skip all this code if it is not defined. i386 is a bit of a corner case, is there even any hardware still available for 32 bit that has AVX512 support? How does the other sse/avx code get around this?

r-devulap commented:
Thanks for taking a look! I am still not entirely sure what's going on, but i got to run now. I will get back to this tomorrow again. Thanks for your help :)

charris commented:
SSE and AVX probably isn't available on 32 bit processors by default. Gcc for the i686 arch assumes Pentium Pro instruction set (1995), later additions need flags. I don't think we support non SSE2 hardware any more, at least on 64 bit processors. There was a discussion at some point. Anyway https://gcc.gnu.org/onlinedocs/gcc-4.8.0/gcc/i386-and-x86_002d64-Options.html shows the options.

r-devulap commented:
Does the CI system test the build of every commit in the PR separately? or just it just build one time that includes all the commits in the PR?

rgommers commented:
> Does the CI system test the build of every commit in the PR separately? or does it just build one time that includes all the commits in the PR?

it builds the most recent commit if all the commits you push at once

rgommers commented:
> SSE and AVX probably isn't available on 32 bit processors by default. Gcc for the i686 arch assumes Pentium Pro instruction set (1995), later additions need flags. I don't think we support non SSE2 hardware any more, at least on 64 bit processors

Correct, we don't support non SSE2 systems anymore. 32-bit Windows binaries are compiled with SSE2, from `msvccompiler.py`:
```
        # msvc9 building for 32 bits requires SSE2 to work around a
        # compiler bug.
        if platform_bits == 32:
            self.compile_options += ['/arch:SSE2']
            self.compile_options_debug += ['/arch:SSE2']
```

Copying [this comment](https://github.com/numpy/numpy/pull/7980#issuecomment-249415479) on the PR that introduced AVX code: 
```
linux always builds generic binaries, this doesn't change this as the unsupported
instructions will never be run if the cpu does not support it.

windows has some special case to build 3 variants with nosse, sse2 and sse3 and 
chooses at install time what to use.
```
That last bit was about installers built with ATLAS, it no longer applies to our current wheels. I'm not sure about how this works on Windows now, but given that the AVX 
```
/*
 * This file is for the definitions of simd vectorized operations.
 *
 * Currently contains sse2 functions that are built on amd64, x32 or
 * non-generic builds (CFLAGS=-march=...)
 * In future it may contain other instruction sets like AVX or NEON detected
 * at runtime in which case it needs to be included indirectly via a file
 * compiled with special options (or use gcc target attributes) so the binary
 * stays portable.
 */


#ifndef __NPY_SIMD_INC
#define __NPY_SIMD_INC

#include "lowlevel_strided_loops.h"
#include "numpy/npy_common.h"
#include "numpy/npy_math.h"
#ifdef NPY_HAVE_SSE2_INTRINSICS      # WILL BE TRUE FOR OUR WHEELS
#include <emmintrin.h>        SSE2
#if !defined(_MSC_VER) || _MSC_VER >= 1600   # WILL BE TRUE FOR OUR WHEELS
#include <immintrin.h>         # AVX
#else
#undef __AVX2__
#undef __AVX512F__
#endif
#endif
```

This is a nice discussion on portability of AVX code on Windows: https://randomascii.wordpress.com/2016/12/05/vc-archavx-option-unsafe-at-any-speed/. It looks like we use the `static inline` trick there, however we don't seem to generate different code for SSE2-only vs AVX available (from `simd.inc.src`):
```
static NPY_INLINE int
run_binary_simd_@kind@_@TYPE@(char **args, npy_intp *dimensions, npy_intp *steps)
{
#if @vector@ && defined NPY_HAVE_SSE2_INTRINSICS
    @type@ * ip1 = (@type@ *)args[0];
    @type@ * ip2 = (@type@ *)args[1];
    @type@ * op = (@type@ *)args[2];
    npy_intp n = dimensions[0];
#if defined __AVX512F__
    const npy_intp vector_size_bytes = 64;
#elif defined __AVX2__
    const npy_intp vector_size_bytes = 32;
#else
    const npy_intp vector_size_bytes = 32;
#endif
```

It would be very useful to document better how this actually works at the moment. We don't use a `/arch:AVX` flag for building wheels, and the MSVC docs say:
```
__AVX__ Defined as 1 when the /arch:AVX or /arch:AVX2 compiler options are set and the compiler target is x86 or x64. Otherwise, undefined.
...
```
So today we require SSE2, and not AVX on Windows. On Linux we do ship with AVX enabled, because there we compile with GCC and it produces portable binaries.

Searching for ["illegal instruction" Windows](https://github.com/numpy/numpy/issues?utf8=%E2%9C%93&q=is%3Aissue+%22illegal+instruction%22+windows) gives only a couple of issues, one that was fixed and the others that were OpenBLAS or VM related. No regular report ever of AVX instructions sneaking into our Windows wheels.

Hope this helps; additions and corrections very welcome!

rgommers commented:
Re 32-bit Linux: there are a number of issues related to VM use where detected CPU instructions are wrong. On CI could work around that perhaps by explicitly disabling AVX512? (haven't thought about that suggestion very hard -  whatever works I'd say, 32-bit performance optimizations aren't very interesting).

mattip commented:
The remaining failures are only on linux32 (which does not have any sse/avx support) and when setting `NPY_RELAXED_STRIDES_CHECKING=0` on linux64. Windows passes CI tests.

r-devulap commented:
> The remaining failures are only on linux32 (which does not have any sse/avx support) and when setting `NPY_RELAXED_STRIDES_CHECKING=0` on linux64. Windows passes CI tests.

The linux64 bit failure is because the CI test is using gcc-4.8. There are several problems with this gcc-4.8: 

1) Use of avx2 intrinsic within a function does not work with just the attribute `__attribute__((target ("avx2")))`. It will explicitly require `-mavx2` flag while compiling. 
2) The `__builtin_cpu_supports` function in gcc does not support avx512f. This too fails at compile time. 

`error: Parameter to builtin not valid: avx512f
         return __builtin_cpu_supports("avx512f") && os_avx512_support();
`

tylerjereddy commented:
That 64-bit linux failure is using Atlas for the linalg backend, so I believe that means we don't have a runtime dependency on the 4.x series gcc / gfortran toolchain libs.

Before bumping the gcc version I guess some thought is warranted about whether it is acceptable to break with the older version, or maybe use preprocessor guards like `#if GCC_VERSION > 40800`

r-devulap commented:
My next question was going to be, is it acceptable to use pre-processor guard checking for GCC version?

rgommers commented:
> My next question was going to be, is it acceptable to use pre-processor guard checking for GCC version?

that seems fine to me. GCC 4.8.x is still common, so we should not drop it.

Commit:
BUG: Adding macro HAVE_ATTRIBUTE_TARGET_@ISA@_WITH_INTRINSICS

1) use __builtin_cpu_supports("avx512f") only for gcc ver >= 5

2) Introduced two new macro's:
HAVE_ATTRIBUTE_TARGET_@ISA@_WITH_INTRINSICS for ensuring compiler can
compile functions that use intrinsics and are compiled with avx2/avx512f
attributes

r-devulap commented:
Fixed the problem with gcc-4.8 build. 2 more failures left, both of which seem to be related to a 32-bit version.

tylerjereddy commented:
Yeah, that's a 32-bit Linux Docker container, probably on a 64-bit machine provided by Azure. Looks like NumPy builds just fine, but `import numpy` is broken.

`_multiarray_umath.cpython-36m-i386-linux-gnu.so: undefined symbol: AVX512F_exp_FLOAT`

Matti will correct me if I'm wrong, but that `.so` likely has link-ins from OpenBLAS, which was pre-built in a 32-bit scenario with different gcc compiler flags & runtime dependencies on a small number of older gfortran libs. OpenBLAS has its own AVX512 stuff too I suppose.

That may not be relevant, but mentioning just in case.

Commit:
BUG: Fixing incomplete guards for @ISA@_exp/log_FLOAT functions

Added a missing NPY_HAVE_SSE2_INTRINSICS guard

r-devulap commented:
> Yeah, that's a 32-bit Linux Docker container, probably on a 64-bit machine provided by Azure. Looks like NumPy builds just fine, but `import numpy` is broken.
> 
> `_multiarray_umath.cpython-36m-i386-linux-gnu.so: undefined symbol: AVX512F_exp_FLOAT`
> 
> Matti will correct me if I'm wrong, but that `.so` likely has link-ins from OpenBLAS, which was pre-built in a 32-bit scenario with different gcc compiler flags & runtime dependencies on a small number of older gfortran libs. OpenBLAS has its own AVX512 stuff too I suppose.
> 
> That may not be relevant, but mentioning just in case.

That was actually a bug in my code, missing pre-processor guards. It only failed on gcc-5 on a 32-bit ubuntu caught it. But all the tests pass now :)

mattip commented:
All tests now passing. This needs a release note in `doc/release/1.17.0-notes.rst.

r-devulap commented:
Any update on this patch?

mattip commented:
There still is a merge conflict with the release notes

mattip commented:
@juliantaylor would you like to take a look?

r-devulap commented:
oh, sorry. I assumed only people with write access can resolve it. I will take care of it.

r-devulap commented:
> There still is a merge conflict with the release notes

took care of the conflict.

r-devulap commented:
I was resolving merge conflict on the release doc and seems like two tests failed with "The install of mingw was NOT successful." I think this has nothing to do with my patch.

tylerjereddy commented:
Looks like a fluke connection failure--I've restarted the job for you

r-devulap commented:
> Looks like a fluke connection failure--I've restarted the job for you

thank you, they all pass now. any word on merging this patch yet?

mattip commented:
Do we have any tests already like the ULP test you describe? It would it be nice to add the ULP test code as a c function to `numpy/core/src/umath/_umath_tests.c.src` and a test to call the function for a vector of a few choice float32 values in `numpy/core/tests/test_ufunc.py`. I think it would be too slow to check the entire range.

At some point we could add "correct value" testing for numpy math functions, much as python [tests cmath](https://github.com/python/cpython/blob/v3.7.3/Lib/test/test_cmath.py#L359) using a file with [expected results](https://github.com/python/cpython/blob/v3.7.3/Lib/test/cmath_testcases.txt) generated by [MPFR](http://www.mpfr.org)

r-devulap commented:
Could you please elaborate on the purpose of such a test? Different algorithms for transcendental functions will not necessarily produce the same exact output, so we can't check for a "correct value". Perhaps we can check if they are within certain acceptable threshold, but that raises two questions: (1) what threshold to use? and (2) which float32 values should we validate in the test?

mattip commented:
> (1) what threshold to use? and (2) which float32 values should we validate in the test?

The threshold is TBD, for instance in the test I linked to `log` is given a larger tolerance than the other tests. The values used to test cmath are the ones known to be problematic: subnormals, near cutoffs in the algorithms, with mantissas near 0 or near the max, -0, ...

It is a lot of work to create the "expected" result. I started an exchange [a few years ago](https://mail.python.org/pipermail/python-dev/2012-September/121645.html) on how to generate the file for 64-bit complex numbers and 32 bit floats, maybe today it would be easier.

r-devulap commented:
sounds good. I am almost done with a another patch vectorizing sine and cosine. Would you mind if I submit the ULP test as part of that patch and we can merge this as is?

Commit:
DOC: adding release notes for 1.17.0

r-devulap commented:
Hi, just checking in on the status of the patch. Let me know if there is anything else you need from me.

mattip commented:
Thanks @r-devulap

seberg commented:
I am not really worried, but I stumbled on this looking up recent speed improvements in NumPy. The speedups here probably mean a slightly larger error on the functions (compared to non-vectorized/fastmath versions)? At what point do we worry about such precision changes?

mattip commented:
xref #13515

r-devulap commented:
@seberg for comparison against libm (scalar) and libmvec ( glibc's vector implementation):

Function | libm max ULP| libmvec max ULP | this patch max ULP
-- | -- | -- | --
sine | 1.19 | 2.47 | 1.49
cosine | 1.19 | 2.39 | 1.49
exp | 0.51 | 2.93 | 2.52
log | 0.51 | 3.96 | 3.83
